File: wjsdncl/141_linked_list_cycle.js
/**
 * 32m15s
 *
 * 시간 복잡도: O(n)
 *
 * 찾아보니 토끼와 거북이 알고리즘으로 풀 수 있는 문제였습니다.
 *
 * 두 포인터를 이용하여 사이클이 존재하는지 확인합니다.
 * slow는 한 칸씩 fast는 두 칸씩 이동합니다.
 * 만약 slow와 fast가 만나면 사이클이 존재합니다.
 * 만나지 않으면 사이클이 존재하지 않습니다.
 */
var hasCycle = function (head) {
  let slow = head;
  let fast = head;

  // fast와 fast.next가 null이 아닌 동안 반복
  while (fast !== null && fast.next !== null) {
    slow = slow.next; // slow는 한 칸 이동
    fast = fast.next.next; // fast는 두 칸 이동

    // slow와 fast가 만나는 경우 사이클 존재
    if (slow === fast) {
      return true;
    }
  }

  // 반복이 끝날 때까지 만나지 않으면 사이클이 없음
  return false;
};

File: citrussoda/4week/cpp/141_linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode* fast = head;
        ListNode* slow = head;

        while (fast != nullptr && fast -> next != nullptr) {
            slow = slow -> next;
            fast = fast -> next -> next;

            if (slow == fast) return true;
        }

        return false;
    }
};


// 2번째 방법

// 다음 노드를 계속 찾아가면서 노드를 visited 됐는지 확인하며 다음으로 진행한다.
// 반복되지 않는다는 것은 언젠가는 끊긴다는 이야기이므로 while로 다음 노드가 NULL이 될 때까지
// 찾아가 저장한다.

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
// class Solution {
// public:
//     bool hasCycle(ListNode *head) {
//         unordered_set<ListNode*> visited;
//         ListNode* current_node = head;

//         while (current_node != nullptr) {
//             // count가 항상 빠른 것은 아니며 큰 차이는 없음. set과 map의 차이인가?
//             if (visited.count(current_node)) return true;

//             visited.insert(current_node);

//             current_node = current_node -> next;
//         }

//         return false;
//     }
// };